### ISSUE TITLE
feat: add organization

### ISSUE THREAD

**Issue by cre8ivejp** [bucketeer-io/bucketeer#704]
# Summary

We plan to introduce the organization object.
The organization encompasses projects and environments.
This feature allows users to add projects and environments freely without system admin operation.

# Spec

## Organization, Project, EnvironmentV2 and AccountV2

```mermaid
erDiagram
organization {
    varchar id
    varchar name
		bool    is_system_admin
}

project {
    varchar id
    varchar name
    varchar organization_id
}

environment_v2 {
    varchar id
    varchar name
    varchar project_id
		varchar organization_id
}

account_v2 {
    varchar email
    varchar name
    varchar avater_image_url
    varchar organization_id
    enum    organization_role(MEMBER_ADMIN_OWNER)
    json    environment_roles(map_env_id_and_envRole)
　　　　　　　
}

organization ||--|{ project : has

project ||--|{ environment_v2 : has

organization ||--|{ account_v2 : has
organization ||--|{ environment_v2 : has
```

- Organization
  - Organization encompasses projects and environments.
- Project
  - Project encompasses environments.
- EnvironmentV2
  - Most Bucketeer resources, such as flags, goals, and experiments, belong to an environment.
- AccountV2
  - Account represents the user who logs in to the management console.
  - Account belongs to the organization.
  - Account is unique by OrganizationID and Email address.
    - You can create accounts in multiple Organizations with the same email address, but each is treated as a different object.
  - Role field expresses the permissions of the Account.
  - The role is discussed below.

## Account's role and AccountV2's role

With this change, the Account object will be replaced by the AccountV2 object.
The AccountV2 object represents a different range than the old Account object Role.

- old: Account.Role
  - This change will make it obsolete.
  - Account.Role has 3 types: viewer, editor, and owner.
- new: AccountV2.Role.Environment and AccountV2.Role.Organization
  - Create two types of roles at the Environment and Organization levels.
 
- AccountV2.Role.Environment
  - represents Environment-level operating permissions.
    - ex.
      - Create, Update, and Delete feature flags in the environment.
      - Create, Update, and Delete goals in the environment.
      - ...
  - 2types
    - VIEWER
    - EDITOR
  
- AccountV2.Role.Organization
  - represents Organization-level operating permissions.
    - ex.
      - Create, Update, and Delete accounts in the organization.
      - Create, Update, and Delete projects in the organization.
      - Create, Update, and Delete environments in the organization.
      - ...
  - 3types
    - MEMBER
      - MEMBER is an Account belonging to Organization, but it cannot operate at Organization level.
    - ADMIN
      - ADMIN can perform essentially all Organization-level operations.
      - However, it is not possible to update, create, or delete the permissions of the Account of the OWNER role.
    - OWNER
      - Like ADMIN, it can perform all Organization-level operations.
      - There must be at least one Account for the OWNER role in an Organization.

(old)Account.Role | (new)AccountV2.Role.Environment | (new)AccountV2.Role.Organization
-- | -- | --
viewer | viewer | member
editor | editor | member
owner | editor | admin

## SystemAdmin Organization, SysemAdmin Account

- The SystemAdmin account is the account that creates the Organization itself.
- In the existing specification, an object called AdminAccounts plays a similar role.
  - They are stored in a DB table called `admin_account`, which is managed separately from the `account` table where normal accounts are stored.
- SystemAdmin accounts (new), unlike Admin accounts (old), are not stored in dedicated tables such as the `admin_account` table.
  - Organization has an `is_system_admin` column (field), and an Organization with this value of true is a SystemAdmin Organization.
  - Accounts belonging to the SystemAdmin Organization are treated as System Admin accounts.

- The SystemAdmin account can perform the following four operations
    - Create a new Organization
    - Update Organization information
    - Archive Organization
    - Unarchive Organization

- The SystemAdmin account can log in to all Organizations and all Environments on the management console.
    - You can log in as orgRole: admin, envRole: editor.

## New API

Name | Service | Usage | Description
-- | -- | -- | --
GetOrganization | environment | organizations list page on the management console | organizations list page on the management consoleis implemented from console 3.0. It does not exist yet.
ListOrganizations | environment | organizations list page on the management console | Same as above
CreateOrganization | environment | organizations list page on the management console | Same as above
UpdateOrganization | environment | organizations list page on the management console | Same as above
EnableOrganization | environment | organizations list page on the management console | Same as above
DisableOrganization | environment | organizations list page on the management console | Same as above
ArchiveOrganization | environment | organizations list page on the management console | Same as above
UnarchiveOrganization | environment | organizations list page on the management console | Same as above
ConvertTrialOrganization | environment | organizations list page on the management console | Same as above
GetAccountV2 | account |   |  
GetAccountV2ByEnvironmentID | account | checkRole() in each microservices | checkRole() checks if the user performing the operation has the necessary permissions. GetAccountV2ByEnvironmentID is executed to obtain account information for this purpose.
ListAccountsV2 | account | accounts list page on the management console | 
CreateAccountV2 | account | accounts list page on the management console | 
UpdateAccountV2 | account | accounts list page on the management console | 
EnableAccountV2 | account | accounts list page on the management console | 
DisableAccountV2 | account | accounts list page on the management console | 
DeleteAccountV2 | account | accounts list page on the management console | 
GetMe | account | During console login (mentioned later) | It takes the organization_id selected by the user and the user's email address and returns an object (ConsoleAccount) with the necessary information on the console.https://github.com/bucketeer-io/bucketeer/pull/665/files#diff-5d492a0187472a3bdbe0d8ec7e5854eae381871fb0f17bdaa75fe4c1b7aa9d51R77-R90
GetMyOrganizations | account | During console login (mentioned later) | Get a list of Organization to which the user belongs from the user's email address. email address is obtained from ServiceToken in the Authorization header, so RequestBody is empty.
GetMyOrganizationsByEmail | account | ExchangeToken API and RefleshToken API(auth service) | Like GetMyOrganizations, GetMyOrganizationsByEmail retrieves a list of organizations to which a user belongs from the user's email address. GetMyOrganizations is used from the console, while GetMyOrganizationsByEmail is used for inter-service communication. Since ServiceToken inside the kubernetes cluster is used in inter-service communication, the email address can be specified in RequestBody.

## Existing APIs that are obsolete

Name | Service | Usage | Description
-- | -- | -- | --
CreateAccount | account | accounts list page on the management console | Replaced by AccountV2 API.
EnableAccount | account | accounts list page on the management console | Replaced by AccountV2 API.
DisableAccount | account | accounts list page on the management console | Replaced by AccountV2 API.
ChangeAccountRole | account | accounts list page on the management console | Replaced by AccountV2 API.
GetAccount | account | accounts list page on the management console | Replaced by AccountV2 API.
ListAccounts | account | accounts list page on the management console | Replaced by AccountV2 API.
CreateAdminAccount | account | admin accounts list page on the management console | The Admin account itself is no longer used, and the admin account page itself is deleted.
EnableAdminAccount | account | admin accounts list page on the management console | Same as above
DisableAdminAccount | account | admin accounts list page on the management console | Same
GetAdminAccount | account | admin accounts list page on the management console | Same as above
ListAdminAccounts | account | admin accounts list page on the management console | Same as above
ConvertAccount | account | admin accounts list page on the management console | Same as above
GetMeV2 | account | During console login | Replaced by GetMe API.
GetMeV2ByEmail | account | ExchangeToken API and RefleshToken API(auth service) | Deleted as no longer in use.
  |   |   |  

## New log-in flow on the management console

![image](https://github.com/bucketeer-io/bucketeer/assets/2755429/e66a9ecf-c513-45b6-b7d5-7ef6b12a42e8)

**Comment by kentakozuka** [comment]
| resource | env viewer | env editor | org member | org admin | org owner | sysadmin|
| --- | --- | --- | --- | --- | --- | --- |
| **Environmental** | | | | | | |
| Feature | view | edit | edit | edit | edit | edit (should be view) |
| Goal | view | edit | edit | edit | edit | edit (should be view) |
| Experiment | view | edit | edit | edit | edit | edit (should be view) |
| Feature | view | edit | edit | edit | edit | edit (should be view) |
| UserSegment | view | edit | edit | edit | edit | edit (should be view) |
| API key | view | edit | view | edit | edit | edit (should be view) |
| Push | - | - | view | edit | edit | edit (should be view) |
| Notification | - | - | view | edit | edit | edit (should be view) |
| Env account | view | view | view | edit | edit | edit (should be view) |
| **Organzational** | | | | | | |
| Environment | - | - | view | edit | edit | edit (should be view) |
| Project | - | - | view | edit | edit | edit (should be view) |
| Org account | - | - | view | edit | edit | edit |
| Organization | - | - | - | - | - | edit |

**Comment by kentakozuka** [comment]
IMHO, system admin shouldn't have access to organizational/environmental resources by default because they contains credential information in their own project.

@cre8ivejp wdyt?

**Comment by cre8ivejp** [comment]
> IMHO, system admin shouldn't have access to organizational/environmental resources by default because they contains credential information in their own project.
> 
> I implemented this concept in this branch #793
> 
> @cre8ivejp wdyt?

Agree. But as we spoke, unfortunately, to do this, we would need to refactor all the internal services because they use the system admin flag to make requests between services. The impact of this would be too big. So it's better to finish the current first, and then we think how to improve it.

**Comment by kentakozuka** [comment]
done
